File: source/ff.resource.editor/xaml/designer/dialog_base.cs
using System.Windows;
using System.Windows.Controls;

namespace editor
{
    public class dialog_base : ContentControl
    {
        public static readonly DependencyProperty TitleProperty = DependencyProperty.Register("Title", typeof(string), typeof(dialog_base));
        public string Title
        {
            get => (string)GetValue(TitleProperty);
            set => SetValue(TitleProperty, value);
        }

        public static readonly DependencyProperty FooterProperty = DependencyProperty.Register("Footer", typeof(UIElement), typeof(dialog_base));
        public UIElement Footer
        {
            get => (UIElement)GetValue(FooterProperty);
            set => SetValue(FooterProperty, value);
        }

        public override void OnApplyTemplate()
        {
            base.OnApplyTemplate();

            if (this.GetTemplateChild("PART_CloseButton") is Button closeButton)
            {
                closeButton.Click += CloseButton_Click;
            }
        }

        private void CloseButton_Click(object sender, RoutedEventArgs e)
        {
        }
    }
}

File: source/ff.resource.editor/source/ui/dialog_base.cpp
#include "pch.h"
#include "source/ui/dialog_base.h"

const Noesis::DependencyProperty* editor::dialog_base::title_property{};
const Noesis::DependencyProperty* editor::dialog_base::footer_property{};

NS_IMPLEMENT_REFLECTION(editor::dialog_base, "editor.dialog_base")
{
    Noesis::DependencyData* data = NsMeta<Noesis::DependencyData>(Noesis::TypeOf<SelfClass>());
    data->RegisterProperty<Noesis::String>(editor::dialog_base::title_property, "Title", Noesis::PropertyMetadata::Create(Noesis::String()));
    data->RegisterProperty<Noesis::Ptr<Noesis::UIElement>>(editor::dialog_base::footer_property, "Footer", Noesis::PropertyMetadata::Create<Noesis::Ptr<Noesis::UIElement>>({}));
}

void editor::dialog_base::OnApplyTemplate()
{
    Noesis::Button* close_button = Noesis::DynamicCast<Noesis::Button*>(this->GetTemplateChild("PART_CloseButton"));
    if (close_button)
    {
        // close_button->Click() += 
    }
}
